src/libxrpl/protocol/STAmount.cpp

```diff
@@ -748,9 +686,18 @@ STAmount::canonicalize()
         {
             Number num(
                 mIsNegative ? -mValue : mValue, mOffset, Number::unchecked{});
-            XRPAmount xrp{num};
-            mIsNegative = xrp.drops() < 0;
-            mValue = mIsNegative ? -xrp.drops() : xrp.drops();
+            if (native())
+            {
+                XRPAmount xrp{num};
+                mIsNegative = xrp.drops() < 0;
+                mValue = mIsNegative ? -xrp.drops() : xrp.drops();
+            }
+            else
+            {
+                MPTAmount c{num};
+                mIsNegative = c.value() < 0;
+                mValue = mIsNegative ? -c.value() : c.value();
+            }
```

Comment:
This code duplication makes me nervous, especially since behind the scenes, the two classes are effectively doing exactly the same thing. What do you think about?
```suggestion
            auto set = [&](auto const& val) {
                mIsNegative = val.value() < 0;
                mValue = mIsNegative ? -val.value() : val.value();
            };
            if (native())
                set(XRPAmount{num});
            else
                set(MPTAmount{num});
```